src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs

```diff
@@ -109,49 +81,47 @@ internal ReturnTypeAndDiagnostics(TypeSymbol returnType, ImmutableArray<Diagnost
         /// </summary>
         internal Binder ScopeBinder { get; }
 
+        public Binder ParameterBinder => _binder;
+
         internal void GetDeclarationDiagnostics(DiagnosticBag addTo)
         {
-            // Force attribute binding for diagnostics
-            foreach (var typeParam in _typeParameters)
+            lock (_lazyLock)
```

Comment:
Does the public API allow binding the same method body from multiple threads? If so, could binding two mutually dependent local functions lead to a deadlock? 

More generally, is the change to use `lock` necessary? Does it handle a case that the `Interlocked.CompareExchange` approach was not handling?